gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/impl/sql/GenericStatement.java

```diff
@@ -759,11 +760,24 @@ private final PreparedStatement prepMinion(
                                           qinfo = qt.computeQueryInfo(qic);
                                           // Only rerouting selects to lead node. Inserts will be handled separately.
                                           // The below should be connection specific.
-                                          if (routeQuery && qinfo != null && qinfo.isSelect() && !isPreparedStatement()) {
-                                            if (SnappyActivation.isColumnTable((DMLQueryInfo)qinfo, false)) {
-                                              return getPreparedStatementForSnappy(true, statementContext, lcc,
-                                                  false, checkCancellation);
+                                          if ((routeQuery && qinfo != null && qinfo.isSelect()
+                                              && !isPreparedStatement() && cc.getExecutionEngine() != ExecutionEngine.STORE)
+                                              && (cc.getExecutionEngine() == ExecutionEngine.SPARK ||
+                                              (new ExecutionEngineArbiter().getExecutionEngine((DMLQueryInfo)qinfo)
```

Comment:
A generic comment to reduce temporary objects. Like a new arbiter for every execution. I don't see any state being maintained. So can the arbiter have just static methods. The state related stuff ( like query info etc is any ways being passed in arguments ).